packages/perennial/contracts/multiinvoker/MultiinvokerRollup.sol

```diff
@@ -0,0 +1,290 @@
+
+                vaultDeposit(depositer, IPerennialVault(vault), amount);
+            } else if (action == 13) { // VAULT_REDEEM
+                address vault; UFixed18 shares;
+                (vault, shares, ptr) = decodeAddressAmount(input, ptr);
+
+                IPerennialVault(vault).redeem(shares, msg.sender);
+            } else if (action == 14) { // VAULT_CLAIM
+                address owner; address vault;
+                (owner, vault, ptr) = decodeAddressAddress(input, ptr);
+
+                IPerennialVault(vault).claim(owner);
+            } else if (action == 15) { // VAULT_WRAP_AND_DEPOSIT 
+                address account; address vault; UFixed18 amount;
+                (account, vault, amount, ptr) = decodeAddressAddressAmount(input, ptr);
+
+                vaultWrapAndDeposit(account, IPerennialVault(vault), amount);
+            }
+        }
+    }
+
+    /// Example Calldata Structure
+    /// let ptr
+    /// [ptr(userLen), ptr+1:userLen(user registry # OR 20 byte address if userLen == 0)] => address user 
+    /// ptr += (userLen OR 20) + 1
+    /// [ptr(prodcutLen), ptr+1:productLen(product registry # OR 20 byte address if prdoctLen == 0)] => address product
+    /// ptr += (prodcutLen OR 20) + 1
+    /// [ptr(amountLen), ptr:amountLen] => uint256 amount 
+
+
+    // ARGUMENT TYPE DECODING //
+
+    function decodeAddressAddressAmount(bytes calldata input, uint ptr) private returns (address addr1, address addr2, UFixed18 amount, uint) {
+        (addr1, ptr) = decodeAddress(input, ptr);
+        (addr2, ptr) = decodeAddress(input, ptr);
+        (amount, ptr) = decodeAmountUFixed18(input, ptr);
+
+        return (addr1, addr2, amount, ptr);
+    }
+
+    function decodeAddressAmount(bytes calldata input, uint ptr) private returns (address addr1, UFixed18 amount, uint) {
+        (addr1, ptr) = decodeAddress(input, ptr);
+        (amount, ptr) = decodeAmountUFixed18(input, ptr);
+
+        return(addr1, UFixed18(amount), ptr);
+    }
+
+    function decodeProductPrograms(bytes calldata input, uint ptr) private returns(address product, uint256[] memory programs, uint) {
+        (product, ptr) = decodeAddress(input, ptr);
+        (programs, ptr) = decodeUintArray(input, ptr);
+
+        return(product, programs, ptr);
+    }
+
+    function decodeAddressAddress(bytes calldata input, uint ptr) private returns(address addr1, address addr2, uint) {
+        (addr1, ptr) = decodeAddress(input, ptr);
+        (addr2, ptr) = decodeAddress(input, ptr);
+
+        return (addr1, addr2, ptr);
+    }
+
+
+    // INDIVIDUAL TYPE DECODING //
+
+     function decodeAmountUFixed18(bytes calldata input, uint ptr) private returns (UFixed18 result, uint) {
+        uint temp;
+        (temp, ptr) = decodeUint(input, ptr);
+
+        return(UFixed18.wrap(temp), ptr);
+    }
+
+    function decodeUint(bytes calldata input, uint ptr) private returns (uint result, uint) {
```

Comment:
can you add checks to all of these that `len` is within an acceptable range (`<32` for uints, `<20` for address, etc). there might be some weird stuff you could do by passing in bad data that way.

Reply:
for decode uint yeah should always be < 32 will add. address is probably covered under that as well bc non '00' length of address is just the uint cache index

Reply:
yeah I think just checking the <32 when decoding is sufficient, was mainly worried about reading/writing into additional slots unintentionally